azure_sdk/lib/azure_sdk/client.rb

```diff
@@ -0,0 +1,537 @@
+require 'azure_mgmt_datalake_analytics'
+require 'azure_mgmt_datalake_store'
+require 'azure_mgmt_devtestlabs'
+require 'azure_mgmt_dns'
+require 'azure_mgmt_features'
+require 'azure_mgmt_graph'
+require 'azure_mgmt_iot_hub'
+require 'azure_mgmt_key_vault'
+require 'azure_mgmt_locks'
+require 'azure_mgmt_logic'
+require 'azure_mgmt_machine_learning'
+require 'azure_mgmt_media_services'
+require 'azure_mgmt_mobile_engagement'
+require 'azure_mgmt_network'
+require 'azure_mgmt_notification_hubs'
+require 'azure_mgmt_policy'
+require 'azure_mgmt_powerbi_embedded'
+require 'azure_mgmt_redis'
+require 'azure_mgmt_resources'
+require 'azure_mgmt_scheduler'
+require 'azure_mgmt_search'
+require 'azure_mgmt_server_management'
+require 'azure_mgmt_service_bus'
+require 'azure_mgmt_sql'
+require 'azure_mgmt_storage'
+require 'azure_mgmt_subscriptions'
+require 'azure_mgmt_traffic_manager'
+require 'azure_mgmt_web'
+
+module Azure::ARM
+  class Client
+    include Azure::ARM::Configurable
+
+    #
+    # Initializes a new instance of the Client class.
+    # @param options [Hash] hash of client options.
+    #
+    def initialize(options = {})
+      if options.is_a?(Hash) && options.length == 0
+        @options = setup_options
+      else
+        @options = options
+      end
+
+      reset!(options)
+    end
+
+    #
+    # Check if this client is configured with the same options.
+    # @param opts [Hash] hash of client options.
+    # @return [Boolean] True if client is configured with same set of options, False otherwise.
+    #
+    def same_options?(opts)
+      opts.hash == @options.hash
+    end
+
+    #
+    # Check if this client is configured with the some options.
+    # @return [Hash] Hash of client configured options if any or empty Hash.
+    #
+    def options
+      @options ||= {}
+    end
+
+    #
+    # Creates a new instance of the MsRest::ServiceClientCredentials class if does not exists.
+    # @return [MsRest::ServiceClientCredentials] Azure ARM credentials to authorize HTTP requests made by the service client.
+    #
+    def credentials
+      if @credentials.nil?
+        self.active_directory_settings ||= Azure::ARM::Default.active_directory_settings
+        puts self.active_directory_settings.authentication_endpoint
+        @credentials = MsRest::TokenCredentials.new(
+            MsRestAzure::ApplicationTokenProvider.new(
+                self.tenant_id, self.client_id, self.client_secret, self.active_directory_settings))
+      end
+      @credentials
+    end
+
+    #
+    # Creates a new instance of the Azure::ARM::Authorization::AuthorizationManagementClient class if does not exists.
+    # @param base_url [String] the base URI of the service.
+    # @param options [Array] filters to be applied to the HTTP requests.
+    # @return [Azure::ARM::Authorization::AuthorizationManagementClient] Azure ARM Authorization Management Client.
+    #
+    def authorization(base_url = nil, options = nil)
```

Comment:
@devigned That's the killer idea. Ruby Meta programming rocks :) 

It'd appreciate if you can spare sometime to review this PR. Thanks!